CodingStyleGuide.md

```diff
@@ -0,0 +1,106 @@
+except for deviations noted here.
+* Follow [Julia contribution guidelines](https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md#general-formatting-guidelines-for-julia-code-contributions), notably its line length
+limit.
+* Follow [Julia guidelines for docstrings](https://docs.julialang.org/en/v1/manual/documentation/index.html).
+* Follow [JuMP coding standards](http://www.juliaopt.org/JuMP.jl/dev/style),
+including its deviations from the Julia style guide.  In particular, note its policies on
+  * [whitespace](http://www.juliaopt.org/JuMP.jl/dev/style/#Whitespace-1)
+  * [return statements](http://www.juliaopt.org/JuMP.jl/dev/style/#Return-statements-1)
+  * [variable names](http://www.juliaopt.org/JuMP.jl/dev/style/#Use-of-underscores-within-names-1).
+* Read [The Zen of Python](https://www.python.org/dev/peps/pep-0020).
+
+## Code Organization
+* Import standard modules, then 3rd-party modules, then yours. Include a blank
+line between each group.
+
+### Modules:  TODO
+
+## Comments
+* Use comments to describe non-obvious or non-trivial aspects of code.
+Describe why something was done but not how.  The "how" should be apparent from
+the code itself.
+* Use complete sentences and proper grammar.
+
+## Constructors
+* Per guidance from Julia documentation, use inner constructors to enforce
+restrictions on parameters or to allow construction of self-referential
+objects.
+Use outer constructors to provide default values or to perform customization.
+* Document the reason why the outer constructor is different.
+* Note that the compiler will provide a default constructor with all struct
+members if no inner constructor is defined.
+* When creating a constructor use "function Foo()" instead of "Foo() = ..."
+One exception is the case where one file has all single-line functions.
+
+## Exceptions
+* Use exceptions for unexpected errors and not for normal error handling.
+  * Detection of an unsupported data format from a user should likely throw
+an exception and terminate the application.
+  * Do not use try/catch to handle retrieving a potentially-missing key from a
+dictionary.
+* Use @assert statements to guard against programming errors. Do not use them
+after detecting bad user input. Note that they may be compiled out in release
+builds.
+
+## Globals
+* Global constants should use UPPER_CASE and be declared const.
+* If global variables are needed, prefix them with "g_".
+* Don't use magic numbers. Instead, define const globals or Enums (Julia
+@enum).
+
+## Macros
+TODO:  JuMP says to use ```@macro(x)``` instead of ```@macro x```.  Do we care?
+
+## One-line Conditionals
+Instead of writing a complete if statement the Julia code base uses this idiom
+frequently:  ```<cond> && <statement>```
+[Example](https://docs.julialang.org/en/v1.0/manual/control-flow/#Short-Circuit-Evaluation-1):
+>
+    function fact(n::Int)
+       n >= 0 || error("n must be non-negative")
+       n == 0 && return 1
+       n * fact(n-1)
```

Comment:
I think using shortcutting `&&` is acceptable, whereas shortcutting `||` is not very readable. As a general rule though I'd avoid both.